### PR TITLE
perf(app): virtualize command list on run detail page

### LINKED ISSUE #9217: Command Item dom performance

### PR DESCRIPTION
**Pull request by b-cooper**
# Overview

In order to make long protocols usable in the new UI, we should only render a small window of the
entire commands list to cover what users are actually able to see in the viewport. Previously large
protocols would gum up the rendering process. This should allow protocols of all sizes to behave like
protocols with only 100 steps in terms of virtual DOM performance. This also adds numbers to each command item which is very useful for visually anchoring users of long protocols.

Closes #9217

# Changelog

- window the Command list by only rendering 100 items at a time, each window overlaps by 50 items, and when scrolling, the window will transition once a user is half of the viewport height away from the window threshold
- add total step count to `CommandList` and individual step numbers to each `CommandItem` in the run log
- make sure command items that have succeeded hare always the same height
- when the `CommandList` mounts, we will automatically set the window to the one that contains the current item, and then scroll so that the current item is at the top of the viewport.

# Review requests

- run a smaller sized protocol, the commands should render as expected in the commands list
- run a very large (>1000 step) protocol, the commands should render as expected and the app should not seize up. E.g. cancelling or pausing a protocol should not be delayed after click.

# Risk assessment
medium, this does touch a lot of the run detail page

### CODE REVIEW

File: api-client/src/runs/types.ts [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-export const RUN_STATUS_SUCCEEDED: 'succeeded' = 'succeeded'
+export const RUN_STATUS_IDLE = 'idle' as const
+export const RUN_STATUS_RUNNING = 'running' as const
+export const RUN_STATUS_PAUSE_REQUESTED = 'pause-requested' as const
+export const RUN_STATUS_PAUSED = 'paused'
+export const RUN_STATUS_STOP_REQUESTED = 'stop-requested' as const
+export const RUN_STATUS_STOPPED = 'stopped' as const
+export const RUN_STATUS_FAILED = 'failed' as const
+export const RUN_STATUS_FINISHING = 'finishing' as const
+export const RUN_STATUS_SUCCEEDED = 'succeeded' as const
```
Comment (shlokamin) [inline api-client/src/runs/types.ts:11]: i find it really annoying we have to do this, I wish TS would do this out of the box, and tbh i don't understand why it doesn't. a string literal can just be interpreted as a subset of a string

File: app/src/organisms/RunDetails/CommandList.tsx [hunk 2]
```diff
@@ -130,164 +146,259 @@ export function CommandList(): JSX.Element | null {
...
 
+  const windowFirstIndex = WINDOW_OVERLAP * windowIndex
+  const commandWindow = currentCommandList.slice(
+    windowFirstIndex,
+    windowFirstIndex + WINDOW_SIZE
+  )
+  const isFirstWindow = windowIndex === 0
+  const isFinalWindow =
+    currentCommandList.length - 1 <= windowFirstIndex + WINDOW_SIZE &&
+    currentCommandList.length - 1 >= windowFirstIndex
```
Comment (shlokamin) [inline app/src/organisms/RunDetails/CommandList.tsx]: don't totally follow this logic, what's going on here?

Comment (shlokamin) [reply app/src/organisms/RunDetails/CommandList.tsx]: I'd think `isFinalWindow` could be determined by doing something like this: 

```js
const numWindows = currentCommandList/WINDOW_SIZE 
const isFinalWindow = windowIndex  === numWindows
```

File: app/src/organisms/RunDetails/CommandList.tsx [hunk 3]
```diff
@@ -130,164 +146,259 @@ export function CommandList(): JSX.Element | null {
...
+    command =>
+      command.runCommandSummary == null ||
+      command.runCommandSummary.status === 'running' ||
+      command.runCommandSummary.status === 'failed' ||
+      command.runCommandSummary.status === 'queued'
+  )
+  const windowIndexWithCurrentItem = Math.floor(
+    Math.max(currentItemIndex - (WINDOW_SIZE - WINDOW_OVERLAP), 0) /
+      WINDOW_OVERLAP
+  )
```
Comment (shlokamin) [inline app/src/organisms/RunDetails/CommandList.tsx]: is this the window index of the current command item? if so can it be renamed to `currentItemWindowIndex` or `windowIndexOfCurrentItem`?

File: app/src/organisms/RunDetails/CommandList.tsx [hunk 4]
```diff
@@ -130,164 +146,259 @@ export function CommandList(): JSX.Element | null {
       : [...postPlayRunCommands]
   }
 
+  const windowFirstIndex = WINDOW_OVERLAP * windowIndex
```
Comment (shlokamin) [inline app/src/organisms/RunDetails/CommandList.tsx]: `windowFirstCommandIndex`

File: app/src/organisms/RunDetails/CommandList.tsx (hunk 2 above)
Comment (shlokamin) [inline app/src/organisms/RunDetails/CommandList.tsx]: do we need this?

File: app/src/organisms/RunDetails/CommandList.tsx [hunk 5]
```diff
@@ -15,35 +17,53 @@ import {
...
-  ] = React.useState<boolean>(false)
   const protocolData: ProtocolFile<{}> | null = useProtocolDetails()
     .protocolData
   const { runRecord } = useCurrentProtocolRun()
   const runStatus = useRunStatus()
+  const listInnerRef = React.useRef<HTMLDivElement>(null)
+  const currentItemRef = React.useRef<HTMLDivElement>(null)
   const runDataCommands = runRecord?.data.commands
+  const [windowIndex, setWindowIndex] = React.useState<number>(0)
+  const [isJumpingToCurrent, setIsJumpingToCurrent] = React.useState<boolean>(
```
Comment (shlokamin) [inline app/src/organisms/RunDetails/CommandList.tsx]: `setIsInitiallyJumpingToCurrent`

**Review by shlokamin — APPROVED** [review]
LGTM! I say let's merge and then come back and abstract the virtualization logic into a hook

File: app/src/organisms/RunDetails/CommandList.tsx (hunk 2 above)
Comment (b-cooper) [reply app/src/organisms/RunDetails/CommandList.tsx]: The window size is 100, but each window overlaps with the previous and next windows by 50. This means that every item falls within two windows technically. 

In order to tell if the current rendered window is the final window, we first:

- get the final command's index in the overall list `currentCommandList.length - 1`
- we compare it to the index of the first command that is rendered in the current window
  - if it is less than or equal to the last index in the current window (`windowFirstIndex + WINDOW_SIZE`)
  - and it's greater than or equal to the first index in the current window
  - then we know that the current window is the final window

Comment (b-cooper) [reply app/src/organisms/RunDetails/CommandList.tsx]: you're right I think we can just check the first condition

**Pull request merged**